Linelist merge

aspina7 opened the pull request:
Starting a new pr to work on function for merging linelists based on dictionaries. 

@alicelouisejane , things we need to discuss: 
- standardising use of "non_template" in dictionary (e.g. i think you put it for some vars like patcurrent_status) 
- removing all NAs in new name and making sure they are tagged with non_template 
- where stuff is and how to approach function (have put some notes in script but just brief thoughts happy to discuss other plans...)

aspina7 committed:
adding variable dictionary

aspina7 committed:
initiating notes for linelist_merge func

alicelouisejane committed:
Updated linelist_diactionary

alicelouisejane committed:
linelist mereg first commit

alicelouisejane committed:
On to cleaning values of the variables of interets. Creating own dictionary from unique values from each varaibale of interest

alicelouisejane committed:
updates

alicelouisejane committed:
updates

alicelouisejane committed:
tested on first 18 files cleaning sheets. In process of creating dictionary needs to be created to clean final merge

alicelouisejane committed:
Works to clean files and merge together. Tested over  ALL files in tamayis country line list files.  cleaning dictionary in process

alicelouisejane committed:
test clean commit

Merge branch 'linelist_merge' of github.com:R4IDSR/covidmonitor into linelist_merge

# Conflicts:
#	R/linelist_merge.R

alicelouisejane committed:
Cleaned merging function and upto date dictionary

alicelouisejane committed:
couple of edits in final merging function

alicelouisejane committed:
First commit of linelist clean

alicelouisejane committed:
edits to merge function

alicelouisejane committed:
stylistic edits to merge function

aspina7 commented:
@alicelouisejane  in rstudio if you go tools > global options > code > display  and then tick the show margin box (the default margin column is 80). 
Then you get a line in your scripts pane - if you make sure your code stays within that line (i.e. break up long lines of code) then is easier to read/review

aspina7 commented on R/linelist_clean.R:
```diff
@@ -0,0 +1,256 @@
+library(dplyr)
+library(data.table)
```
best to avoid data.table - different syntax and they wont be familiar with it - so for front end cleaning stuff stick to tidyverse or base.

aspina7 replied:
if there is something you arent sure how to do without data.table flag it and will take a look - most things are replicable at this point and the speed increases dont really justify the density with the data size we are working at. There is also the [dtplyr ](https://dtplyr.tidyverse.org/) package if interested.

aspina7 commented on R/linelist_clean.R:
```diff
@@ -0,0 +1,256 @@
+
+#patinfo_occus
+#list of words/ strings associated with healthcare
+occupation<-rio::import(here::here("inst/","Cleaning_dict_alice.xlsx"), which=1) %>% select(patinfo_occus)
```
the dictionaries in inst are installed when the package from github - so they are in programme files (or applications on a mac) under the package folder. Use syst.file function as have done in the [merge_linelist function](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-cfb1eba336b200fc57bfa3888fdcb8678770651acdbe2319e87105afc47c1634R15). This way you avoid having to have multiple excel files together in the same project folder each time you use the function.

aspina7 replied:
also please move all the datasets reading in to the top of script

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+
+    #if ISO codes is not in initilsed list of interest skip this file, this could be defined already in template dictionary ?
+    if(!(iso %in% isotomerge)){
+      print(paste(iso,"skip"))
```
use warnings rather than print

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+
+    #if ISO codes is not in initilsed list of interest skip this file, this could be defined already in template dictionary ?
+    if(!(iso %in% isotomerge)){
+      print(paste(iso,"skip"))
```
use warnings and errors rather than print 
to just give messages about stuff you have done but dont want the function to stop use warning, e.g. [here](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-5581348288db36d823a2992aea1e71eb9b0847865fa1e4668c905c5a91734fbdR287). 
If want to make function stop and show an error message then use stop, e.g. [here](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-5581348288db36d823a2992aea1e71eb9b0847865fa1e4668c905c5a91734fbdR42)

aspina7 replied:
this applies to everywhere that you use print - but just gonna flag it here cus lazy.

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+  var_dict <- rio::import(
+    here::here("inst","extdata", "linelist_dictionary.xlsx"),
+    which = 1
+  ) %>% as_tibble() %>%
```
dont think you need to say as_tibble here

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    here::here("inst","extdata", "linelist_dictionary.xlsx"),
+    which = 1
+  ) %>% as_tibble() %>%
+    linelist::clean_variable_names() %>% mutate_at("old_var_r", gsub,pattern ="\\s+$", replacement ="" ) %>%
```
use janitor::clean_names rather than linelist. Linelist I dont think is on cran yet and is not stable anyway, janitor much better. 
Also any functions that you use outside of base you need to list up top in the @importFrom bit

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    which = 1
+  ) %>% as_tibble() %>%
+    linelist::clean_variable_names() %>% mutate_at("old_var_r", gsub,pattern ="\\s+$", replacement ="" ) %>%
+    mutate_at("old_var_r", gsub,pattern ="\\s+", replacement =" ") %>% # cleaning old variable names from unwanted regex patterns
```
dont use pipes (%>%) in functions, makes debugging harder. 
Add separate lines

alicelouisejane replied:
removed all except in date part (since you're doing the date function)

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    linelist::clean_variable_names() %>% mutate_at("old_var_r", gsub,pattern ="\\s+$", replacement ="" ) %>%
+    mutate_at("old_var_r", gsub,pattern ="\\s+", replacement =" ") %>% # cleaning old variable names from unwanted regex patterns
+    mutate_at("old_var_r", gsub,pattern ="\\s+|[[:punct:]]", replacement ="_") %>% mutate_at(c("old_var_r","new_var"),tolower) %>%
+    mutate_at(c("old_var_r","new_var"),stringi::stri_trans_general,"Latin-ASCII")# cleaning old variable names replacing all spaces and punctuation with . (as this is dont on the variable names upon loading files)
```
add stringi to imports

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+
+#initialise list of ISO codes from the countries we are interested in merging
+#this is based off Francks files
+isotomerge<- c("BFA","CIV","COD","COG","DZA","GIN","KEN","LBR","MOZ","MUS","NAM","NER","RWA","SEN","SLE","STP","SWZ","SYC","TCD","UGA","ZWE")
```
add this as an argument to the function - e.g. how have done [here](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-5581348288db36d823a2992aea1e71eb9b0847865fa1e4668c905c5a91734fbdR27) and then subset files how have done [here](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-5581348288db36d823a2992aea1e71eb9b0847865fa1e4668c905c5a91734fbdR46). 
"AFRO" will read in all files (default) and if specify certain iso codes then will only read in those.

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    # load in file
+    #files that are xlsb (CIV) cannot be open (easily on mac) for ease resave the file in an xlsx format, this can be edited to use the RODBC package which i think works on windows
+    if (grepl(files[f],pattern="\\.xlsb$", ignore.case=TRUE)){
+      print(paste(iso, "file is in xlsb format, please resave file in xlsx format before proceeding"))
```
use the readxlsb::read_xlsb
syntax similar and output should be too - easier to automate this than hope people will actually read and understand the message

alicelouisejane replied:
read_xlsb is awfully slow and actually never loaded and crashes R. Others found similar here 
[https://stackoverflow.com/questions/28684199/how-to-open-an-xlsb-file-in-r](url)
the RODBC package which is an alternative is for Microsoft OS only so hence why I have done it this way. Cant get it to work on my machine so maybe you could edit this bit and check if works with RODBC  ?

aspina7 replied:
weird - read_xlsb worked fine for me in a diff project .... rodbc definitely works but clunky, will give it a whirl

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+      og_sheet <- rio::import(files[f], which=sheetname, skip=skip)
+    }
+
+    # clean variable names to removed unwated regex patterns, replace all spaces with a ., inorder to match var_dict dictionary
```
janitor::clean_names should do all this for you.

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    #remove all accents from values in dataframe
+    #replace all values in file that should be coded as missing
+    #using gsub and regex pattern matching here as some patterns require to be fixed and others not
+    og_sheet<-og_sheet %>% mutate_all(stringi::stri_trans_general,"Latin-ASCII") %>%
```
i think all the mutate_ functions are going to be deprecated. Better to use across e.g. [here ](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-cb168d009c6e28df2f67926fd089c0b46f5e11b9e80c7caeac4ba1cd7e8a5e8aR53)

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+#must keep this pattern all one line or doesnt work
+
+    # filter variable cleaning dictionary specific to country file loaded
+    var_dict_country<- var_dict %>% filter(country==iso)
```
dont use pipe, just chuck var_dict in there filter(vardict, country... bla)

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    drop_nontemplate_vars<- as.list(var_dict_country$old_var_r[var_dict_country$notes=="non_template" & !is.na(var_dict_country$notes)])
+
+    # drop variables from dataframe which are non-tmeplate
+    if(length(drop_nontemplate_vars)!=0){
```
chuck a warning with which ones dropped with the iso and week too

alicelouisejane replied:
do you mean epi week or just use Sys.date ?

aspina7 replied:
umm think i meant the epiweek variable - like chuck a warning saying "Country X week Y dropped blabla variables"

alicelouisejane replied:
don't think all have the epi week variable... any idea how they calculate it cos I could just do that based of the date

alicelouisejane replied:
have just left this in warning as todays date (probs need it to be date in the file name) can figure out what to put here

aspina7 replied:
they must all have a report date though no? Or unique ID and country?  will take a look at this in the next round...

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    var_dict_country<- var_dict_country[var_dict_country$old_var_r %in% names(og_sheet),]
+
+    # list variables to drop that are non-template
+    drop_nontemplate_vars<- as.list(var_dict_country$old_var_r[var_dict_country$notes=="non_template" & !is.na(var_dict_country$notes)])
```
dont think this needs to be as.list can just be a vector and then use !which instead of -which

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+      #remove following instances from variable names taht need recoding for the next step to leave what will become the recoded value
+      names(recode_sheet)<- gsub(x = names(recode_sheet), pattern = "COVID.|Co19.|comcond.|patsympt.|patinfo.", replacement = "", ignore.case = T)
+
+      #replace all instances of no with missing
```
you could save yourself some density here by using a grepl for the Yes options and then if sum is the number you want

aspina7 replied:
TRUE/FALSE is treated as 1/0

alicelouisejane replied:
some columns that are part of the recode have values that I want to keep still (that are not yes), its an annoying step but I think is only safe way round it

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+      #renames variables with key of what to code to
+      names(recode_sheet)<-paste(names(recode_sheet),variable_to, sep= ".")
+
+      require(data.table)
```
dont require packages in the function - add it to imports. Okay to use data.table here as is backend, but rethink if actually need to this because dependencies and breaking...

alicelouisejane replied:
recoded to not depend on data.table, will double checks works in all instances but seems ok

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+      names(output_sheet) <- with(var_dict_country,new_var[match(names(output_sheet),old_var_r)])
+      #cbind with newly recoded variables
+      output_sheet$id<-rownames(output_sheet)#create an id variable for when merging back to og sheet occurs
+      output_sheet<-merge(output_sheet,recode_sheet,by="id") %>% select(-c(id))
```
pipe

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+      }
+
+      #handle all date variables
+      require(anytime)
```
add to imports

aspina7 replied:
nice package tho! hadnt seen before

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    # Convert to character date for final merge
+    output_sheet$id<-rownames(output_sheet)
+
+    datesnumeric<-output_sheet %>% select(contains("Date", ignore.case = T),id) %>% mutate_all(.funs=gsub,pattern=".",replacement = "/", fixed=T, perl = T) %>%
```
pipes and change to across - but also it looks like anytime should be able to deal with these without having to do all the extra recoding no?

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+      # Convert to numeric and change date format using origin of excel
+      # Convert to character date for final merge
+      datescharacter<-output_sheet %>% select(contains("Date", ignore.case = T), id) %>%
+        mutate_at(vars(contains("Date", ignore.case = T)), lubridate::parse_date_time, orders=c("ymd","Ymd","dmy","dmY","%Y-%m-%d","%y-%m-%d","%d-%m-%y","dBY","ymd HMS","Ymd HMS")) %>%
```
lubridate to imports if cant do with anytime (one or the other must be able to all of this no?)

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+
+       #Natural join of data frames, fills in missing values in one df with values from other if not misssing
+
+      dates<- rquery::natural_join(datesnumeric, datescharacter,
```
equivalent of dplyr full_join no? reduce dependencies

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    if(!("report_date" %in% vars)){
+      output_sheet$report_date<-NA
+      output_sheet$report_date<-as.Date(output_sheet$report_date,origin="1899-12-30")
+      print("report_date")
```
warnings... do a big combined message with all the added/dropped vars, e.g. [here](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-5581348288db36d823a2992aea1e71eb9b0847865fa1e4668c905c5a91734fbdR287)

alicelouisejane replied:
as above

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+      names(output_sheet)<-tolower(names(output_sheet))
+
+  ######
+    #checking that variables of interest exists
```
think the way I did [here ](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-5581348288db36d823a2992aea1e71eb9b0847865fa1e4668c905c5a91734fbdR257) might be shorter?

alicelouisejane replied:
do you mean this for dates or creating missing vars ?

alicelouisejane replied:
made much shorter for creating the missing vars with added warnings etc

aspina7 replied:
ah sorry - for some reason links to line references not working. But yes meant for creating missing vars and chucking warnings.

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    #add cleaned output sheet to a list
+    output[[f]] <-output_sheet
+  }
+    #merge all cleaned sheets into one
```
dplyr bind_rows

alicelouisejane replied:
Reduce is beneficial as doesn't matter about differing column classes. Avoids converting to column classes here and losing data (for example in the age column there is text written which is handled in the clean function)

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+    big_data<-Reduce(function(...) merge(..., all=T), output_fin)
+    big_data<-big_data[order(big_data$country_iso),]
+    # write xlsx file to take on to ext step of cleaning
+    openxlsx::write.xlsx(big_data, here::here("inst",paste0(outputname, Sys.Date(),".xlsx")))
```
dont do to inst - this should be to whatever file they specify in the function arguments

alicelouisejane replied:
changed to rio::export

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+require(dplyr)
+
+  # read in dictionary for renaming variables contry specific sheet
+  var_dict <- rio::import(
```
see comment in cleaning script about importing from system.file. Also move all files read in to top of function for oversight

alicelouisejane replied:
can't get System.file to work for me not sure what I'm doing wrong, have added in where I thought it should be and commented out just so I could run it myself and generate output

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+  # create folder for output
+  base::dir.create(outputdirectory, showWarnings = FALSE)
+
+require(dplyr)
```
dont require in function. put it in the imports above and then build when you are working on bits

aspina7 reviewed (commented):
been through and commented (mostly in the function), you can see all the comments with the lines of code if you click on files changed up top - will go through the cleaning once we finalise the merge function.  lmk if cant see comments or not clear

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,416 @@
+
+      #handle all date variables
+      require(anytime)
+    #handle dates that are numeric. All character dates -> NA.
```
working on a [function ](https://github.com/R4IDSR/covidmonitor/pull/9/files#diff-cb168d009c6e28df2f67926fd089c0b46f5e11b9e80c7caeac4ba1cd7e8a5e8aR56) for dates - message me when you get here

alicelouisejane replied:
left this part as discussed

alicelouisejane committed:
Merge branch 'linelist_merge' of github.com:R4IDSR/covidmonitor into linelist_merge

# Conflicts:
#	R/linelist_merge.R

alicelouisejane committed:
edits to merge from comments

alicelouisejane committed:
added additional "NA" detection

aspina7 commented on R/linelist_merge.R:
```diff
@@ -0,0 +1,353 @@
+      }
+    }
+
+    # handle dates that are numeric. All character dates -> NA.
```
note to self - swap in date function

alicelouisejane committed:
minor edits

alicelouisejane committed:
updated

alicelouisejane committed:
update to dictionary, cleaning dictionary added and cleaning file

alicelouisejane committed:
update and adding dictionaries

Merge branch 'linelist_merge' of github.com:R4IDSR/covidmonitor into linelist_merge

# Conflicts:
#	R/linelist_merge.R

alicelouisejane committed:
first commit of analysis script for who is dying manusript

alicelouisejane committed:
editied mauritius template from TRUE to FALSE

alicelouisejane committed:
final commit to linelist merge branch
